pytorch_lightning/utilities/cli.py

```diff
@@ -0,0 +1,274 @@
+                <https://omni-us.github.io/jsonargparse/#classes-as-type>`_ of the
+                given class.
+            subclass_mode_data: Whether datamodule can be any `subclass
+                <https://omni-us.github.io/jsonargparse/#classes-as-type>`_ of the
+                given class.
+        """
+        self.model_class = model_class
+        self.datamodule_class = datamodule_class
+        self.save_config_callback = save_config_callback
+        self.trainer_class = trainer_class
+        self.trainer_kwargs = {} if trainer_kwargs is None else trainer_kwargs
+        self.subclass_mode_model = subclass_mode_model
+        self.subclass_mode_data = subclass_mode_data
+        self.parser_kwargs = {} if parser_kwargs is None else parser_kwargs
+        self.parser_kwargs.update({
+            'description': description,
+            'env_prefix': env_prefix,
+            'default_env': env_parse
+        })
+
+        self.init_parser()
+        self.add_arguments_to_parser(self.parser)
+        self.add_core_arguments_to_parser()
+        self.before_parse_arguments(self.parser)
+        self.parse_arguments()
+        self.before_instantiate_classes()
+        self.instantiate_classes()
+        self.prepare_fit_kwargs()
+        self.before_fit()
+        self.fit()
+        self.after_fit()
+
+    def init_parser(self):
+        """Method that instantiates the argument parser"""
+        self.parser = LightningArgumentParser(**self.parser_kwargs)
+
+    def add_arguments_to_parser(self, parser: LightningArgumentParser):
+        """Implement to add extra arguments to parser
+
+        Args:
+            parser: The argument parser object to which arguments should be added
+        """
+        pass
+
+    def add_core_arguments_to_parser(self):
+        """Adds arguments from the core classes to the parser"""
+        self.parser.add_trainer_args(self.trainer_class, 'trainer')
+        self.parser.add_module_args(self.model_class, 'model', subclass_mode=self.subclass_mode_model)
+        if self.datamodule_class is not None:
+            self.parser.add_datamodule_args(self.datamodule_class, 'data', subclass_mode=self.subclass_mode_data)
+
+    def before_parse_arguments(self, parser: LightningArgumentParser):
+        """Implement to run some code before parsing arguments
+
+        Args:
+            parser: The argument parser object that will be used to parse
+        """
+        pass
+
+    def parse_arguments(self):
+        """Parses command line arguments and stores it in self.config"""
+        self.config = self.parser.parse_args()
+
+    def before_instantiate_classes(self):
+        """Implement to run some code before instantiating the classes"""
+        pass
+
+    def instantiate_classes(self):
+        """Instantiates the classes using settings from self.config"""
+        self.config_init = self.parser.instantiate_subclasses(self.config)
+        self.instantiate_model()
```

Comment:
I still feel it is better to change the order here, so that datamodule would be instantiated before the model. There are common use cases where the model construction depends on the data used (especially for the input/output layers). I haven't yet found a use case where the datamodule depends on the model. While it is true that the use can still override the whole of instantiate classes as needed, I feel that more boilerplate can be saved if the order is datamodule, then model, then trainer.

Reply:
Done